src/stores/MessagePreviewStore.ts

```diff
@@ -0,0 +1,132 @@
+    private static internalInstance = new MessagePreviewStore();
+
+    private constructor() {
+        super(defaultDispatcher, {});
+    }
+
+    public static get instance(): MessagePreviewStore {
+        return MessagePreviewStore.internalInstance;
+    }
+
+    /**
+     * Gets the pre-translated preview for a given room
+     * @param room The room to get the preview for.
+     * @returns The preview, or null if none present.
+     */
+    public getPreviewForRoom(room: Room): string {
+        if (!room) return null; // invalid room, just return nothing
+
+        // It's faster to do a lookup this way than it is to use Object.keys().includes()
+        const val = this.state[room.roomId];
+        if (val !== null && typeof(val) !== "string") {
+            this.generatePreview(room);
+        }
+
+        return this.state[room.roomId];
+    }
+
+    private generatePreview(room: Room) {
+        const maxEventsBackwards = 50; // any further and we just assume there's nothing important
+
+        const timeline = room.getLiveTimeline();
+        if (!timeline) return; // usually only happens in tests
+        const events = timeline.getEvents();
+
+        for (let i = events.length - 1; i >= 0; i--) {
+            if (i === events.length - maxEventsBackwards) return; // limit reached
+
+            const event = events[i];
+            const preview = this.generatePreviewForEvent(event);
+            if (preview.isPreviewable) {
+                // noinspection JSIgnoredPromiseFromCall - the AsyncStore handles concurrent calls
+                this.updateState({[room.roomId]: preview.preview});
+                return; // break - we found some text
+            }
+        }
+
+        // if we didn't find anything, subscribe ourselves to an update
+        // noinspection JSIgnoredPromiseFromCall - the AsyncStore handles concurrent calls
+        this.updateState({[room.roomId]: null});
+    }
+
+    protected async onAction(payload: ActionPayload) {
+        if (!this.matrixClient) return;
+
+        // TODO: Remove when new room list is made the default
+        if (!RoomListStoreTempProxy.isUsingNewStore()) return;
+
+        if (payload.action === 'MatrixActions.Room.timeline' || payload.action === 'MatrixActions.Event.decrypted') {
+            const event = payload.event; // TODO: Type out the dispatcher
+            if (!Object.keys(this.state).includes(event.getRoomId())) return; // not important
+
+            const preview = this.generatePreviewForEvent(event);
+            if (preview.isPreviewable) {
+                await this.updateState({[event.getRoomId()]: preview.preview});
+                return; // break - we found some text
+            }
+        }
+    }
+
+    private generatePreviewForEvent(event: MatrixEvent): { isPreviewable: boolean, preview: string } {
+        if (PREVIEWABLE_EVENTS.some(p => p.type === event.getType() && p.isState === event.isState())) {
```

Comment:
this feels really weird not being a map or switch/case

Reply:
... how would you do that? It's an array.

Reply:
Right, but why is PREVIEWABLE_EVENTS an array? it makes more sense being an object/map keyed by the type

Reply:
not when you also want to filter on whether or not its a state event. The last thing we need is someone posting `m.room.topic` events to the regular timeline and the room list thinking it's a state event.

Reply:
`if (PREVIEWABLE_EVENTS[event.getType()] && PREVIEWABLE_EVENTS[event.getType()].isState === event.isState()` ?

Reply:
That's awful looking imho :(

Reply:
or like EventTile: https://github.com/matrix-org/matrix-react-sdk/blob/013751e3083147141cb7a1504e84069a2a7b6c70/src/components/views/rooms/EventTile.js#L114

Reply:
Also not great. A single list of things we care about feels more correct, as it's not like we're actually mapping config info onto something. Tiles are a bit different only because it's literally mapping an event type to a component, while here's we're just trying to figure out if it makes the list of events we care about.